website/www/site/content/en/documentation/patterns/shared-class.md

```diff
@@ -0,0 +1,187 @@
+    def __init__(self, shared_handle):
+        self._version = 1
+        self._customers = {}
+        self._shared_handle = shared_handle
+
+    def setup(self):
+        # setup is a good place to initialize transient in-memory resources.
+        self._customer_lookup = self._shared_handle.acquire(self.load_customers)
+
+    def load_customers(self):
+        time.sleep(2)
+        self._customers = gen_customers(version=self._version)
+        return WeakRefDict(self._customers)
+
+    def process(self, element):
+        attr = self._customer_lookup.get(element["customer_id"], {})
+        yield {**element, **attr}
+
+
+with beam.Pipeline() as p:
+    shared_handle = shared.Shared()
+    (
+        p
+        | beam.Create(gen_orders(ts=datetime.now().timestamp()))
+        | beam.ParDo(EnrichOrderFn(shared_handle))
+        | beam.Map(print)
+    )
+
+    logging.getLogger().setLevel(logging.INFO)
+    logging.info("Building pipeline ...")
+{{< /highlight >}}
+
+## Create a cache and update it regularly on a streaming pipeline
+
+As the batch pipeline example, the customer cache is initialized as a dictionary in the `setup` method. To update the cache, it is refreshed when its timestamp is earlier than the defined threshold value of 5 seconds. The refreshment is implemented in the `start_bundle` method. Once the cached is updated, the order records are enriched with updated customer attributes. Note that it is necessary to change the `tag` argument of the `acquire` method to reload the shared object when it gets refreshed. The current timestamp value (`current_ts`) is used in this example.
+
+{{< highlight py >}}
+import random
+import logging
+import time
+from datetime import datetime
+from uuid import uuid4
+
+import apache_beam as beam
+from apache_beam.utils import shared
+from apache_beam.transforms.periodicsequence import PeriodicImpulse
+
+
+def gen_customers(version: int, num_cust: int = 1000):
+    d = dict()
+    for r in range(num_cust):
+        d[r] = {"version": version}
+    d["timestamp"] = datetime.now().timestamp()
+    return d
+
+
+def gen_orders(e: float, num_ord: int = 5, num_cust: int = 1000):
+    orders = [
+        {
+            "order_id": str(uuid4()),
+            "customer_id": random.randrange(1, num_cust),
+            "timestamp": e,
+        }
+        for _ in range(num_ord)
+    ]
+    for o in orders:
+        yield o
+
+
+# wrapper class needed for a dictionary since it does not support weak references.
+class WeakRefDict(dict):
+    pass
+
+
+class EnrichOrderFn(beam.DoFn):
+    def __init__(self, shared_handle):
+        self._max_stale_sec = 5
+        self._version = 1
+        self._customers = {}
+        self._shared_handle = shared_handle
+
+    def setup(self):
+        # setup is a good place to initialize transient in-memory resources.
+        self._customer_lookup = self._shared_handle.acquire(
+            self.load_customers, datetime.now().timestamp()
```

Comment:
Calling acquire with a timestamp based tag in setup is probably a bad idea since this will be consistently changing for each thread which calls setup.

Reply:
Would you consider it is okay if it is changed as

```python
    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers, "init")
```

Reply:
I think using the timestamp bucketized version I mentioned below probably makes more sense (see that thread for explanation of why)

Reply:
Resolved by a separate commit.